enhancements/machine-api/machine-deletion-hooks.md

```diff
@@ -0,0 +1,319 @@
+to synchronise the Etcd data to the replacement Machine.
+
+This could also be used by other components such as Storage to allow them to
+ensure safe detachment of volumes before the Machine is deleted.
+
+### Goals
+
+- Define an initial set of hook points for the deletion phase.
+- Define an initial set and form of related annotations.
+- Define basic expectations for a controller or process that responds to a
+lifecycle hook.
+
+### Non-Goals
+
+- Create an exhaustive list of hooks; we can add more over time.
+- Create new machine phases.
+- Create a mechanism to signal what lifecycle point a machine is at currently.
+- Dictate implementation of controllers that respond to the hooks.
+- Implement ordering in the machine controller.
+- Require anyone to use these hooks for normal machine operations, these are
+strictly optional and for custom integrations only.
+
+## Proposal
+
+- Utilize annotations to implement lifecycle hooks.
+- Each lifecycle point can have 0 or more hooks.
+- Hooks do not enforce ordering.
+- Hooks found during machine reconciliation effectively pause reconciliation
+until all hooks for that lifecycle point are removed from a machine's annotations.
+
+### User Stories
+
+#### Story 1
+(pre-terminate) As an operator, I would like to have the ability to perform
+different actions between the time a Machine is marked deleted in the API and
+the time the Machine is deleted from the infrastructure provider.
+
+For example, when replacing a Control Plane Machine, ensure a new Control
+Plane Machine has been successfully created and joined to the cluster before
+removing the instance of the deleted Machine. This might be useful in case
+there are disruptions during replacement and we need the disk of the existing
+instance to perform some disaster recovery operation.  This will also prevent
+prolonged periods of having one fewer Control Plane host in the event the
+replacement instance does not come up in a timely manner.
+
+#### Story 2
+(pre-drain) As an operator, I want the ability to utilize my own draining
+controller instead of the logic built into the machine controller.  This will
+allow me better flexibility and control over the lifecycle of workloads on each
+node.
+
+### Implementation Details/Notes/Constraints
+
+For each defined lifecycle point, one or more hooks may be applied as an annotation
+to the machine object. These annotations will pause reconciliation of a Machine object
+until all hooks are resolved for that lifecycle point.
+The hooks should be managed by a Hook Implementing Controller or other external
+application, or manually created and removed by an administrator.
+
+#### Lifecycle Points
+
+##### pre-drain
+
+`pre-drain.delete.hook.machine.cluster.x-k8s.io`
+
+Hooks defined at this point will prevent the machine controller from draining a node
+after the machine object has been marked for deletion until the hooks are removed.
```

Comment:
> Again my concern about potentially blocking upgrades because of unexpected annotation it seems like this perhaps should have stricter validation enforcement via CR possibly?

These hooks only come into play when a machine is being deleted. This is not something that would normally happen during an upgrade scenario. Upgrades typically involve MCO which updates the components on the control plane nodes in place and reboots them for the changes to take effect, Machine API doesn't come into this process and neither would this feature.

I can't find another comment about blocking upgrades, have I missed something? 🤔 

> How will we know who added an annotation and why? The Reason for the hook might be very useful in understanding intent.

In the `Annotation Form` section below, we show some examples of what this might look like. The idea is that the name of the hook captures the reason for why the hook exists, and the value captures who created it. I'm hesitant to go down the route of a CR as it seems a lot of extra bulk for not a huge gain (I see the gain as a longer limit on the reason field).

I would expect for the etcd operator we would add the following hook:
```
pre-drain.delete.hook.machine.cluster-api.x-k8s.io/etcd-voting-member: openshift-etcd-operator
```
Would you want to put anything more into the context than this?

Reply:
I only say block upgrade out of an error case. so for example for whatever reason, the hook was never deleted resulting in a block on drain. User for whatever reason does not know this is the case and attempts upgrade. In this case I believe this hook would/could block an upgrade.

Reply:
Just a corner case we would want to provide clear signal on.

Reply:
> User for whatever reason does not know this is the case and attempts upgrade.

Machine API doesn't factor into upgrades so I don't think this would be a case that we need to worry about. During upgrades, MCO will rollout updates to the Nodes, which does its own draining and AFAIK isn't intended to hook into this

The only reason today that Machine API would be involved in an upgrade is if you need to create new capacity to allow the upgrade to go ahead, in which case these hooks wouldn't be involved either.

I think alerting on blocked removal of Machines should be sufficient signal for most cases given this, and we already have alerts like that set up